File: collector/src/dcim_collector/dnstap.py
"""dnstap reader for the collector's DNS observability loop.

CoreDNS's `dnstap` plugin streams every query handled by the auth
pod to a UNIX socket on the shared dns-state volume; this module
listens on that socket and decodes each query just far enough to
pull out the question name + type. The decoded `(name, type)` tuple
is fed back to the collector via a callback, which folds it into a
per-server top-K reservoir the metrics loop ships in its POST.

Why hand-rolled and not `dnstap-py` / `fstrm-py`:
  - dnstap-py last released in 2018, depends on a protobuf release
    line we'd otherwise drag into the collector for one message type.
  - The wire formats we need (fstrm control + data framing, dnstap
    Dnstap message, and the DNS question section) total under 200
    lines of straightforward decoding. Cheaper to maintain than the
    dependency surface of the third-party packages.

Bidirectional fstrm handshake (CoreDNS is the client, we listen):
  reader: <accept>                  writer: <connect>
  reader: <read READY>             writer: <send READY>
  reader: <send ACCEPT(ctype)>     writer: <read ACCEPT>
  reader: <read START>             writer: <send START(ctype)>
       --- data frames flow ---
  reader: <read STOP>              writer: <send STOP>
  reader: <send FINISH>            writer: <read FINISH>
"""

from __future__ import annotations

import asyncio
import os
import struct
from collections.abc import Awaitable, Callable
from pathlib import Path

import structlog

log = structlog.get_logger("collector.dnstap")

# Frame Streams control-frame types per draft-ietf-dnsop-dnstap-09 §3.
_CTL_ACCEPT = 0x01
_CTL_START = 0x02
_CTL_STOP = 0x03
_CTL_READY = 0x04
_CTL_FINISH = 0x05

_CTL_FIELD_CONTENT_TYPE = 0x01
_DNSTAP_CONTENT_TYPE = b"protobuf:dnstap.Dnstap"

# RFC 1035 question-section TYPE values plus the few DNSSEC + service
# types operators are likely to want to surface on the dashboard.
# Numeric fall-through ("17", "65535") covers anything not enumerated
# here so the top-K never drops a sample just because we don't have a
# pretty label for the type.
_RR_TYPES = {
    1: "A", 2: "NS", 5: "CNAME", 6: "SOA", 12: "PTR", 15: "MX",
    16: "TXT", 24: "SIG", 28: "AAAA", 33: "SRV", 35: "NAPTR", 41: "OPT",
    43: "DS", 46: "RRSIG", 47: "NSEC", 48: "DNSKEY", 50: "NSEC3",
    51: "NSEC3PARAM", 52: "TLSA", 65: "HTTPS", 257: "CAA",
}

OnQuery = Callable[[str, str], Awaitable[None] | None]


# --- low-level helpers --------------------------------------------------------


async def _read_exact(reader: asyncio.StreamReader, n: int) -> bytes:
    """Read exactly n bytes or raise EOFError on premature close."""
    buf = b""
    while len(buf) < n:
        chunk = await reader.read(n - len(buf))
        if not chunk:
            raise EOFError
        buf += chunk
    return buf


def _make_control_frame(payload: bytes) -> bytes:
    """Wrap a control payload in fstrm's `0x00000000 escape + length`
    prefix. Used when we send ACCEPT / FINISH back to CoreDNS."""
    return b"\x00\x00\x00\x00" + struct.pack(">I", len(payload)) + payload


def _make_accept_frame() -> bytes:
    """ACCEPT control frame carrying the `protobuf:dnstap.Dnstap`
    content-type field — tells CoreDNS we accept dnstap payloads."""
    field = (
        struct.pack(">I", _CTL_FIELD_CONTENT_TYPE)
        + struct.pack(">I", len(_DNSTAP_CONTENT_TYPE))
        + _DNSTAP_CONTENT_TYPE
    )
    return _make_control_frame(struct.pack(">I", _CTL_ACCEPT) + field)


def _make_finish_frame() -> bytes:
    return _make_control_frame(struct.pack(">I", _CTL_FINISH))


async def _read_frame(reader: asyncio.StreamReader) -> tuple[str, bytes]:
    """Decode the next fstrm frame. Returns `(kind, payload)` where
    kind is `'control'` or `'data'`."""
    length = struct.unpack(">I", await _read_exact(reader, 4))[0]
    if length == 0:
        ctrl_length = struct.unpack(">I", await _read_exact(reader, 4))[0]
        return ("control", await _read_exact(reader, ctrl_length))
    return ("data", await _read_exact(reader, length))


# --- protobuf decoder ---------------------------------------------------------


def _read_varint(data: bytes, pos: int) -> tuple[int, int]:
    """Decode a protobuf varint at `pos`. Returns `(value, new_pos)`."""
    result = 0
    shift = 0
    while pos < len(data):
        byte = data[pos]
        result |= (byte & 0x7F) << shift
        pos += 1
        if not (byte & 0x80):
            return result, pos
        shift += 7
        if shift >= 64:
            raise ValueError("varint too long")
    raise ValueError("varint truncated")


def _skip_field(data: bytes, pos: int, wire_type: int) -> int:
    """Skip a protobuf field we don't care about. Returns new pos."""
    if wire_type == 0:
        _, pos = _read_varint(data, pos)
    elif wire_type == 1:
        pos += 8
    elif wire_type == 2:
        length, pos = _read_varint(data, pos)
        pos += length
    elif wire_type == 5:
        pos += 4
    else:
        raise ValueError(f"unknown wire type {wire_type}")
    return pos


def _find_length_delimited(
    data: bytes, target_field: int,
) -> bytes | None:
    """Walk a protobuf message looking for a length-delimited field
    matching `target_field`. Returns the inner bytes or None.

    Used twice on each frame: once on the outer `Dnstap` message
    (target field 14 = `message`), once on the inner `Message`
    struct (target field 10 = `query_message`)."""
    pos = 0
    while pos < len(data):
        tag, pos = _read_varint(data, pos)
        field_num = tag >> 3
        wire_type = tag & 0x07
        if field_num == target_field and wire_type == 2:
            length, pos = _read_varint(data, pos)
            return data[pos:pos + length]
        pos = _skip_field(data, pos, wire_type)
    return None


def _extract_query_wire(frame: bytes) -> bytes | None:
    """Pull the raw DNS-wire-format query bytes out of a dnstap
    frame. Returns None when the frame is a response-only message or
    is malformed enough that we shouldn't trust the rest."""
    try:
        inner = _find_length_delimited(frame, target_field=14)
        if inner is None:
            return None
        return _find_length_delimited(inner, target_field=10)
    except (ValueError, IndexError):
        return None


# --- DNS wire format ---------------------------------------------------------


def _decode_dns_question(wire: bytes) -> tuple[str | None, str | None]:
    """Parse just the question section of a DNS message. Returns
    `(qname, qtype_label)` or `(None, None)` on malformed input.

    We deliberately don't use dnspython for this — the question
    section is fixed-shape (12-byte header + length-prefixed labels
    + 4 bytes type/class) and pulling in a 100KB dependency for
    that one decode would be silly."""
    if len(wire) < 12:
        return None, None
    pos = 12  # skip the DNS header
    labels: list[str] = []
    while pos < len(wire):
        length = wire[pos]
        pos += 1
        if length == 0:
            break
        if length & 0xC0:
            # Compression pointers don't appear in the question
            # section per RFC 1035 §4.1.4 — treat as malformed.
            return None, None
        if pos + length > len(wire):
            return None, None
        labels.append(wire[pos:pos + length].decode("ascii", "replace"))
        pos += length
    if pos + 4 > len(wire):
        return None, None
    qtype = struct.unpack(">H", wire[pos:pos + 2])[0]
    qname = (".".join(labels) + ".").lower() if labels else "."
    return qname, _RR_TYPES.get(qtype, str(qtype))


# --- per-connection handler --------------------------------------------------


async def _read_until_ready(reader: asyncio.StreamReader) -> bool:
    """Pump frames until we see a READY control frame. Returns True
    on success, False on EOF or unexpected payload."""
    kind, payload = await _read_frame(reader)
    if kind != "control" or len(payload) < 4:
        return False
    return struct.unpack(">I", payload[:4])[0] == _CTL_READY


async def _await_start(reader: asyncio.StreamReader) -> bool:
    """Read frames until START arrives. CoreDNS sends START with the
    same content_type field we already verified on its READY frame,
    so we don't re-check it here — we just need to advance past it
    before data frames begin."""
    kind, payload = await _read_frame(reader)
    if kind != "control" or len(payload) < 4:
        return False
    return struct.unpack(">I", payload[:4])[0] == _CTL_START


async def _dispatch_query(frame: bytes, on_query: OnQuery) -> None:
    """Decode one data frame and invoke the consumer callback. Any
    decode failure is silently dropped — a single torn frame
    shouldn't bring the dnstap loop down."""
    query_wire = _extract_query_wire(frame)
    if query_wire is None:
        return
    qname, qtype = _decode_dns_question(query_wire)
    if qname is None or qtype is None:
        return
    result = on_query(qname, qtype)
    if asyncio.iscoroutine(result):
        await result


async def _data_loop(
    reader: asyncio.StreamReader,
    writer: asyncio.StreamWriter,
    on_query: OnQuery,
) -> None:
    """Read data frames until STOP arrives, then send FINISH back."""
    while True:
        kind, payload = await _read_frame(reader)
        if kind == "control":
            ctl_type = struct.unpack(">I", payload[:4])[0]
            if ctl_type == _CTL_STOP:
                writer.write(_make_finish_frame())
                await writer.drain()
                return
            # Unknown control frames mid-stream — ignore and keep
            # reading. The fstrm spec allows future extensions.
            continue
        await _dispatch_query(payload, on_query)


async def _handle_client(
    reader: asyncio.StreamReader,
    writer: asyncio.StreamWriter,
    on_query: OnQuery,
) -> None:
    """One CoreDNS dnstap connection. Walks the bidirectional fstrm
    handshake then loops reading data frames until STOP/EOF."""
    peer = "unix"
    try:
        if not await _read_until_ready(reader):
            return
        writer.write(_make_accept_frame())
        await writer.drain()
        if not await _await_start(reader):
            return
        await _data_loop(reader, writer, on_query)
    except (EOFError, ConnectionResetError, asyncio.IncompleteReadError):
        pass
    except Exception as e:  # noqa: BLE001
        log.warning("dnstap_client_error", peer=peer, err=str(e))
    finally:
        writer.close()
        try:
            await writer.wait_closed()
        except Exception:  # noqa: BLE001
            pass


# --- public entrypoint -------------------------------------------------------


async def serve_dnstap(socket_path: str, on_query: OnQuery) -> None:
    """Run a UNIX-socket fstrm server forever. CoreDNS connects to
    us as a client; we invoke `on_query(name, type)` for every
    decoded DNS query.

    Removes a stale socket file from a previous run before binding
    — important when the collector restarts because Linux leaves the
    inode behind and `bind()` would EADDRINUSE otherwise."""
    Path(socket_path).parent.mkdir(parents=True, exist_ok=True)
    try:
        Path(socket_path).unlink()
    except FileNotFoundError:
        pass
    server = await asyncio.start_unix_server(
        lambda r, w: _handle_client(r, w, on_query),
        path=socket_path,
    )
    # Lock the socket down to owner-only. Both the collector and the
    # resolver (CoreDNS / Hickory) must run as the same UID — the
    # site-dns compose + k8s manifests do this by running both as root
    # in the shared dns-state volume. Operators running the resolver
    # as a different UID need to align it with the collector's UID
    # (or chown the socket post-bind via their orchestration).
    try:
        os.chmod(socket_path, 0o600)
    except OSError:
        pass
    log.info("dnstap_server_start", socket=socket_path)
    async with server:
        await server.serve_forever()

File: packages/badger/internal/dnstap/dnstap.go
// Package dnstap is the Go port of collector/src/dcim_collector/dnstap.py.
// CoreDNS streams every query handled by an auth pod to a UNIX socket
// via its `dnstap` plugin; this package listens on the socket, walks
// fstrm's bidirectional handshake, and decodes each data frame just
// far enough to pull out the DNS question name + type. The decoded
// (name, type) tuple is fed to a caller-supplied OnQuery callback,
// which the metrics loop uses to fold queries into a per-server
// top-K reservoir.
//
// Why hand-rolled rather than using github.com/dnstap/golang-dnstap:
//   - The upstream lib pulls in golang/protobuf v1 + four indirect
//     deps for one wire message. We need ~250 lines of fstrm framing
//     + ~60 lines of manual protobuf decoding to skip the dep entirely.
//   - Parity with the Python collector's hand-rolled approach keeps
//     the two implementations easy to reason about side-by-side.
//
// Bidirectional fstrm handshake (CoreDNS connects to us as the client):
//
//	reader: <accept>                  writer: <connect>
//	reader: <read READY>             writer: <send READY>
//	reader: <send ACCEPT(ctype)>     writer: <read ACCEPT>
//	reader: <read START>             writer: <send START(ctype)>
//	     --- data frames flow ---
//	reader: <read STOP>              writer: <send STOP>
//	reader: <send FINISH>            writer: <read FINISH>
package dnstap

import (
	"bufio"
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// OnQuery is invoked for every decoded DNS query. Implementations must
// be cheap and non-blocking (or hand off to a worker); the dnstap loop
// runs single-threaded per connection.
type OnQuery func(name, qtype string)

// Frame Streams control-frame types per draft-ietf-dnsop-dnstap-09 §3.
const (
	ctlAccept = 0x01
	ctlStart  = 0x02
	ctlStop   = 0x03
	ctlReady  = 0x04
	ctlFinish = 0x05

	ctlFieldContentType = 0x01
)

var dnstapContentType = []byte("protobuf:dnstap.Dnstap")

// RFC 1035 question TYPE values + the few DNSSEC + service types we
// want pretty labels for. Anything else falls through to the numeric
// string so the top-K never drops a sample for lack of a label.
var rrTypes = map[uint16]string{
	1: "A", 2: "NS", 5: "CNAME", 6: "SOA", 12: "PTR", 15: "MX",
	16: "TXT", 24: "SIG", 28: "AAAA", 33: "SRV", 35: "NAPTR", 41: "OPT",
	43: "DS", 46: "RRSIG", 47: "NSEC", 48: "DNSKEY", 50: "NSEC3",
	51: "NSEC3PARAM", 52: "TLSA", 65: "HTTPS", 257: "CAA",
}

// Serve listens on socketPath until ctx is cancelled. CoreDNS connects
// to us; each connection is handled in its own goroutine. A stale
// socket file is removed before bind — important after a kill -9.
func Serve(ctx context.Context, socketPath string, onQuery OnQuery, log *slog.Logger) error {
	if err := os.MkdirAll(filepath.Dir(socketPath), 0o755); err != nil {
		return fmt.Errorf("mkdir socket parent: %w", err)
	}
	_ = os.Remove(socketPath)
	l, err := net.Listen("unix", socketPath)
	if err != nil {
		return fmt.Errorf("listen unix %s: %w", socketPath, err)
	}
	// Loosen perms so CoreDNS (which may run as a non-root UID) can
	// connect; the shared volume is the security boundary.
	_ = os.Chmod(socketPath, 0o660)

	go func() {
		<-ctx.Done()
		_ = l.Close()
	}()

	log.Info("dnstap_server_start", "socket", socketPath)
	for {
		conn, err := l.Accept()
		if err != nil {
			if ctx.Err() != nil {
				return nil
			}
			return fmt.Errorf("accept: %w", err)
		}
		go handleClient(conn, onQuery, log)
	}
}

func handleClient(conn net.Conn, onQuery OnQuery, log *slog.Logger) {
	defer conn.Close()
	br := bufio.NewReaderSize(conn, 65536)

	// 1. Read READY from CoreDNS.
	if !readControl(br, ctlReady) {
		return
	}
	// 2. Send ACCEPT with our advertised content type.
	if _, err := conn.Write(makeAcceptFrame()); err != nil {
		return
	}
	// 3. Read START from CoreDNS, then begin draining data frames.
	if !readControl(br, ctlStart) {
		return
	}

	for {
		kind, payload, err := readFrame(br)
		if err != nil {
			if !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
				log.Debug("dnstap_read_failed", "err", err)
			}
			return
		}
		if kind == frameControl {
			if len(payload) < 4 {
				continue
			}
			ctl := binary.BigEndian.Uint32(payload[:4])
			if ctl == ctlStop {
				_, _ = conn.Write(makeFinishFrame())
				return
			}
			// Spec allows future control frames mid-stream — ignore.
			continue
		}
		// Data frame: dnstap-encoded message.
		dispatch(payload, onQuery)
	}
}

// frame-kind sentinels make the read loop self-documenting.
type frameKind int

const (
	frameControl frameKind = iota
	frameData
)

// readFrame decodes one fstrm frame. Layout:
//
//	[4-byte BE length]
//	  if length == 0: escape — followed by [4-byte ctrl length][ctrl payload]
//	  else:           [length bytes of data payload]
func readFrame(r io.Reader) (frameKind, []byte, error) {
	var lenBuf [4]byte
	if _, err := io.ReadFull(r, lenBuf[:]); err != nil {
		return 0, nil, err
	}
	length := binary.BigEndian.Uint32(lenBuf[:])
	if length == 0 {
		// control-frame escape
		if _, err := io.ReadFull(r, lenBuf[:]); err != nil {
			return 0, nil, err
		}
		ctl := binary.BigEndian.Uint32(lenBuf[:])
		buf := make([]byte, ctl)
		if _, err := io.ReadFull(r, buf); err != nil {
			return 0, nil, err
		}
		return frameControl, buf, nil
	}
	buf := make([]byte, length)
	if _, err := io.ReadFull(r, buf); err != nil {
		return 0, nil, err
	}
	return frameData, buf, nil
}

// readControl pumps frames until a control frame matching `want`
// arrives. Returns false on EOF or mismatched payload.
func readControl(r io.Reader, want uint32) bool {
	kind, payload, err := readFrame(r)
	if err != nil || kind != frameControl || len(payload) < 4 {
		return false
	}
	return binary.BigEndian.Uint32(payload[:4]) == want
}

func makeControlFrame(payload []byte) []byte {
	out := make([]byte, 8+len(payload))
	// escape prefix
	binary.BigEndian.PutUint32(out[0:4], 0)
	binary.BigEndian.PutUint32(out[4:8], uint32(len(payload)))
	copy(out[8:], payload)
	return out
}

func makeAcceptFrame() []byte {
	field := make([]byte, 8+len(dnstapContentType))
	binary.BigEndian.PutUint32(field[0:4], ctlFieldContentType)
	binary.BigEndian.PutUint32(field[4:8], uint32(len(dnstapContentType)))
	copy(field[8:], dnstapContentType)

	payload := make([]byte, 4+len(field))
	binary.BigEndian.PutUint32(payload[0:4], ctlAccept)
	copy(payload[4:], field)
	return makeControlFrame(payload)
}

func makeFinishFrame() []byte {
	var p [4]byte
	binary.BigEndian.PutUint32(p[:], ctlFinish)
	return makeControlFrame(p[:])
}

// dispatch decodes one dnstap data frame and invokes onQuery if the
// frame contains a question section we could read. Decode failures
// drop silently — one torn frame shouldn't stop the dnstap loop.
func dispatch(frame []byte, onQuery OnQuery) {
	inner := findLengthDelimited(frame, 14) // Dnstap.message
	if inner == nil {
		return
	}
	wire := findLengthDelimited(inner, 10) // Message.query_message
	if wire == nil {
		return
	}
	name, qtype, ok := decodeDNSQuestion(wire)
	if !ok {
		return
	}
	onQuery(name, qtype)
}

// --- minimal protobuf decoder -----------------------------------------------

func readVarint(b []byte, pos int) (uint64, int, error) {
	var result uint64
	var shift uint
	for pos < len(b) {
		bt := b[pos]
		result |= uint64(bt&0x7F) << shift
		pos++
		if bt&0x80 == 0 {
			return result, pos, nil
		}
		shift += 7
		if shift >= 64 {
			return 0, pos, errors.New("varint too long")
		}
	}
	return 0, pos, errors.New("varint truncated")
}

func skipField(b []byte, pos int, wireType uint64) (int, error) {
	switch wireType {
	case 0:
		_, p, err := readVarint(b, pos)
		return p, err
	case 1:
		return pos + 8, nil
	case 2:
		length, p, err := readVarint(b, pos)
		if err != nil {
			return pos, err
		}
		return p + int(length), nil
	case 5:
		return pos + 4, nil
	default:
		return pos, fmt.Errorf("unknown wire type %d", wireType)
	}
}

// findLengthDelimited walks a protobuf message and returns the inner
// bytes of the first length-delimited (wire-type 2) field matching
// targetField. Used twice per dnstap frame.
func findLengthDelimited(data []byte, targetField int) []byte {
	pos := 0
	for pos < len(data) {
		tag, np, err := readVarint(data, pos)
		if err != nil {
			return nil
		}
		pos = np
		fieldNum := int(tag >> 3)
		wireType := tag & 0x07
		if fieldNum == targetField && wireType == 2 {
			length, p, err := readVarint(data, pos)
			if err != nil {
				return nil
			}
			pos = p
			if pos+int(length) > len(data) {
				return nil
			}
			return data[pos : pos+int(length)]
		}
		newPos, err := skipField(data, pos, wireType)
		if err != nil || newPos <= pos {
			return nil
		}
		pos = newPos
	}
	return nil
}

// --- DNS wire decoder --------------------------------------------------------

// decodeDNSQuestion parses just the question section of a DNS message.
// Returns (qname, qtype_label, ok). RFC 1035 §4.1.4 forbids compression
// pointers in the question section, so we treat any 0xC0+ length byte
// as malformed.
func decodeDNSQuestion(wire []byte) (string, string, bool) {
	if len(wire) < 12 {
		return "", "", false
	}
	pos := 12 // skip DNS header
	var labels []string
	for pos < len(wire) {
		length := int(wire[pos])
		pos++
		if length == 0 {
			break
		}
		if length&0xC0 != 0 {
			return "", "", false
		}
		if pos+length > len(wire) {
			return "", "", false
		}
		labels = append(labels, string(wire[pos:pos+length]))
		pos += length
	}
	if pos+4 > len(wire) {
		return "", "", false
	}
	qtype := binary.BigEndian.Uint16(wire[pos : pos+2])
	name := "."
	if len(labels) > 0 {
		name = strings.ToLower(strings.Join(labels, ".") + ".")
	}
	label, ok := rrTypes[qtype]
	if !ok {
		label = strconv.Itoa(int(qtype))
	}
	return name, label, true
}
